src/app/clusters/tls-certificate-management-server/tls-certificate-management-server.cpp

```diff
@@ -0,0 +1,416 @@
+        break;
+    case FindRootCertificate::Id:
+        CommandHandlerInterface::HandleCommand<FindRootCertificate::DecodableType>(
+            ctx, [this](HandlerContext & innerCtx, const auto & req) { HandleFindRootCertificate(innerCtx, req); });
+        break;
+    case LookupRootCertificate::Id:
+        CommandHandlerInterface::HandleCommand<LookupRootCertificate::DecodableType>(
+            ctx, [this](HandlerContext & innerCtx, const auto & req) { HandleLookupRootCertificate(innerCtx, req); });
+        break;
+    case RemoveRootCertificate::Id:
+        CommandHandlerInterface::HandleCommand<RemoveRootCertificate::DecodableType>(
+            ctx, [this](HandlerContext & innerCtx, const auto & req) { HandleRemoveRootCertificate(innerCtx, req); });
+        break;
+    case TLSClientCSR::Id:
+        CommandHandlerInterface::HandleCommand<TLSClientCSR::DecodableType>(
+            ctx, [this](HandlerContext & innerCtx, const auto & req) { HandleGenerateClientCsr(innerCtx, req); });
+        break;
+    case ProvisionClientCertificate::Id:
+        CommandHandlerInterface::HandleCommand<ProvisionClientCertificate::DecodableType>(
+            ctx, [this](HandlerContext & innerCtx, const auto & req) { HandleProvisionClientCertificate(innerCtx, req); });
+        break;
+    case FindClientCertificate::Id:
+        CommandHandlerInterface::HandleCommand<FindClientCertificate::DecodableType>(
+            ctx, [this](HandlerContext & innerCtx, const auto & req) { HandleFindClientCertificate(innerCtx, req); });
+        break;
+    case LookupClientCertificate::Id:
+        CommandHandlerInterface::HandleCommand<LookupClientCertificate::DecodableType>(
+            ctx, [this](HandlerContext & innerCtx, const auto & req) { HandleLookupClientCertificate(innerCtx, req); });
+        break;
+    case RemoveClientCertificate::Id:
+        CommandHandlerInterface::HandleCommand<RemoveClientCertificate::DecodableType>(
+            ctx, [this](HandlerContext & innerCtx, const auto & req) { HandleRemoveClientCertificate(innerCtx, req); });
+        break;
+    }
+}
+
+void TlsCertificateManagementServer::HandleProvisionRootCertificate(HandlerContext & ctx,
+                                                                    const ProvisionRootCertificate::DecodableType & req)
+{
+    ChipLogDetail(Zcl, "TlsCertificateManagement: ProvisionRootCertificate");
+
+    VerifyOrReturn(req.certificate.size() <= kSpecMaxCertBytes,
+                   ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::ConstraintError));
+    VerifyOrReturn(req.caid.IsNull() || req.caid.Value() <= kMaxRootCertId,
+                   ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::ConstraintError));
+
+    ProvisionRootCertificateResponse::Type response;
+    auto status = mDelegate.ProvisionRootCert(ctx.mRequestPath.mEndpointId, ctx.mCommandHandler.GetAccessingFabricIndex(), req,
+                                              response.caid);
```

Comment:
Before forwarding to the delegate, the provisioning handler should also check if the same cert already exists which will return `CertificateAlreadyInstalled` error.

> If the CAID is null:
...
 If the Node already has the certificate installed:
   Fail the command with a cluster-specific status code of CertificateAlreadyInstalled, and
end processing with no other side effects.

We could possibly load the certs here from the delegate (as also done for the read encoding above) and check for existence, or return this cluster specific error from the delegate. But I think the former may be the preferred approach just so we have it at a common server cluster handling spot and don't rely on cert tests to validate if various apps implementing the delegate did the right thing.

Reply:
You actually bring up an issue with the spec, filed https://github.com/CHIP-Specifications/connectedhomeip-spec/issues/11898

Updating code to match what the spec will say after that fix